### PR TITLE
Add claim name validation

### PR DESCRIPTION
**Pull request by nessiLahav**
### What does this PR do?
Add claim name validation

#### Change log
- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [ ] The changes in this PR do not affect the Conjur API

### CODE REVIEW

File: app/domain/logs.rb [hunk 1]
```diff
@@ -590,6 +590,20 @@ module AuthnJwt
         code: "CONJ00116D"
       )
```
Comment (nessiLahav) [inline app/domain/logs.rb:592]: @shulifink @hilagross Please review

File: app/domain/errors.rb [hunk 2]
```diff
@@ -484,6 +484,22 @@ module AuthnJwt
         code: "CONJ00102E"
       )
```
Comment (nessiLahav) [inline app/domain/errors.rb:486]: @shulifink @hilagross Please review

File: app/domain/authentication/authn_jwt/consts.rb [hunk 3]
```diff
@@ -20,11 +20,17 @@ module AuthnJwt
     EXP_CLAIM_NAME = "exp"
     NBF_CLAIM_NAME = "nbf"
     IAT_CLAIM_NAME = "iat"
-    SUPPORTED_ALGORITHMS = %w[RS256 RS384 RS512].freeze
+    JTI_CLAIM_NAME = "jti"
+    RSA_ALGORITHMS = %w[RS256 RS384 RS512].freeze
+    ECDSA_ALGORITHMS = %w[ES256 ES384 ES512].freeze
```
Comment (benoaviram) [inline app/domain/authentication/authn_jwt/consts.rb:25]: @nessiLahav  did you rebase? this is a change that Sasha did on purpose..

File: app/domain/authentication/authn_jwt/consts.rb [hunk 4]
```diff
@@ -20,11 +20,17 @@ module AuthnJwt
...
+    SUPPORTED_ALGORITHMS = (RSA_ALGORITHMS + ECDSA_ALGORITHMS)
     CACHE_REFRESHES_PER_INTERVAL = 10
     CACHE_RATE_LIMIT_INTERVAL = 300
     CACHE_MAX_CONCURRENT_REQUESTS = 3
     MANDATORY_CLAIMS = [EXP_CLAIM_NAME].freeze
     OPTIONAL_CLAIMS = [ISS_CLAIM_NAME, NBF_CLAIM_NAME, IAT_CLAIM_NAME].freeze
+    VALID_CLAIM_NAME_REGEX = /^[a-zA-Z|$|_][a-zA-Z|$|_|0-9]*$/ # starts with letter $ or _ can contains digits
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/consts.rb:32]: @sashaCher please review

File: app/domain/authentication/authn_jwt/consts.rb [hunk 5]
```diff
@@ -20,11 +20,15 @@ module AuthnJwt
...
+    JTI_CLAIM_NAME = "jti"
     SUPPORTED_ALGORITHMS = %w[RS256 RS384 RS512].freeze
     CACHE_REFRESHES_PER_INTERVAL = 10
     CACHE_RATE_LIMIT_INTERVAL = 300
     CACHE_MAX_CONCURRENT_REQUESTS = 3
     MANDATORY_CLAIMS = [EXP_CLAIM_NAME].freeze
     OPTIONAL_CLAIMS = [ISS_CLAIM_NAME, NBF_CLAIM_NAME, IAT_CLAIM_NAME].freeze
+    VALID_CLAIM_NAME_REGEX = /^[a-zA-Z|$|_][a-zA-Z|$|_|0-9]*$/ # starts with letter $ or _ can contains digits
+    MANDATORY_CLAIMS_RESOURCE_NAME = "mandatory-claims"
+    MANDATORY_CLAIMS_DENY_LIST = [ISS_CLAIM_NAME, EXP_CLAIM_NAME, NBF_CLAIM_NAME, IAT_CLAIM_NAME, JTI_CLAIM_NAME].freeze
```
Comment (benoaviram) [inline app/domain/authentication/authn_jwt/consts.rb:32]: @nessiLahav what should we do with other public claims like : auth_time (https://www.iana.org/assignments/jwt/jwt.xhtml) we will allow it?

Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/consts.rb:32]: why do we work with deny list instead of allow list for the allowed claims?

File: app/domain/errors.rb [hunk 6]
```diff
@@ -484,6 +484,22 @@ module AuthnJwt
...
+
+      FailedToValidateClaimForbiddenClaimName = ::Util::TrackableErrorClass.new(
+        msg: "Failed to validate claim, claim name '{0-claim-name}' is forbidden, " \
+             " please validate all claims defined in '{1-mandatory-claims-variable-name}' " \
+             "are matching '{2-regex}' regular expression.",
+        code: "CONJ00104E"
+      )
+
+      FailedToValidateClaimClaimNameInDenyList  = ::Util::TrackableErrorClass.new(
+        msg: "Failed to validate claim, claim name '{0-claim-name}' in deny list '{1-deny-list}'",
```
Comment (Avi-Eyal) [inline app/domain/errors.rb:500]:
```suggestion
        msg: "Failed to validate claim, claim name '{0-claim-name}' is in deny list '{1-deny-list}'",
```

File: app/domain/authentication/authn_jwt/consts.rb (hunk 3 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/consts.rb]: Yes i fixed it thanks

File: app/domain/authentication/authn_jwt/consts.rb (hunk 5 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/consts.rb:32]: because this is custom claim so you can configure what ever you want except of the JWT reserved claims

File: app/domain/errors.rb (hunk 6 above)
Comment (nessiLahav) [reply app/domain/errors.rb]: thanks

File: app/domain/authentication/authn_jwt/consts.rb (hunk 5 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/consts.rb:32]: @benoaviram good question

@sashaCher @hilagross WDYT?

**Comment by tzheleznyak** [comment]
Address code climate. 
What we decide to won't fix add doc string explaining why

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        dependencies: {
+          regexp_class: Regexp,
+          deny_claims_list_value: [],
+          logger: Rails.logger
+        },
+        inputs: %i[claim_name]
+      ) do
+        def call
+          @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatingClaimName.new(@claim_name))
+          validate_claim_name_exists
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:15]: If we use regex to validate the claim? It will also validate its not empty ?
or we want different error for empty claims?

Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:15]: i prefer a different error because you should not call this command class with nil or empty value

**Review by tzheleznyak — DISMISSED** [review]
Left 2 small comments.

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
+module Authentication
+  module AuthnJwt
+    module InputValidation
```
[the remark this answers is not in the document]
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:3]: Wont fix: Redundant comment

[the remark this answers is not in the document]
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:3]: Wont fix: Redundant comment

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        
+        def claim_name_regex
+          @claim_name_regex ||= Regexp.new(VALID_CLAIM_NAME_REGEX)
+        end
+        
+        def validate_claim_is_allowed
+          @logger.debug(LogMessages::Authentication::AuthnJwt::MandatoryClaimsDenyListValue.new(@deny_claims_list_value))
+          return if @deny_claims_list_value.blank?
+
+          if @deny_claims_list_value.include?(@claim_name)
```
[the remark this answers is not in the document]
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:45]: Wont fix: The current style is more clear

File: app/domain/authentication/authn_jwt/consts.rb (hunk 5 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/consts.rb:32]: No limitations from my point of view. May be we need to put some general disclaimer in doc that time valued claims could not be annotations, if it's not obvious.

Comment (sashaCher) [reply app/domain/authentication/authn_jwt/consts.rb:32]: @nessiLahav Please add `aud` to the `DENY_LIST`.

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        def validate_claim_name_exists
+          raise Errors::Authentication::AuthnJwt::FailedToValidateClaimMissingClaimName if @claim_name.blank?
+        end
+
+        def validate_claim_name_value
+          return if claim_name_regex.match?(@claim_name)
+
+          raise Errors::Authentication::AuthnJwt::FailedToValidateClaimForbiddenClaimName.new(
+            @claim_name,
+            MANDATORY_CLAIMS_RESOURCE_NAME,
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:32]: `MANDATORY_CLAIMS_RESOURCE_NAME` looks unrelated to this great class and in general to the PR.
Why do we couple this pure logic to the resource name.
I think that we should keep it resource independent.
Will Tamir use the same class for mapping validation?

File: app/domain/errors.rb [hunk 11]
```diff
@@ -484,6 +484,22 @@ module AuthnJwt
...
+        msg: "Failed to validate claim, claim name '{0-claim-name}' is forbidden, " \
+             " please validate all claims defined in '{1-mandatory-claims-variable-name}' " \
+             "are matching '{2-regex}' regular expression.",
+        code: "CONJ00104E"
+      )
+
+      FailedToValidateClaimClaimNameInDenyList  = ::Util::TrackableErrorClass.new(
+        msg: "Failed to validate claim, claim name '{0-claim-name}' is in deny list '{1-deny-list}'",
+        code: "CONJ00105E"
+      )
```
Comment (hilagross) [inline app/domain/errors.rb:502]: 1. I'm not sure we the phrase 'claim name' is correct, I think it should be either claim or claim key.
2. denylist in one word -> see [this](https://ca-il-confluence.il.cyber-ark.com/display/rndp/%5BSpike%5D+Rename+and+offensive+of+conjur#id-[Spike]Renameandoffensiveofconjur-Specificchanges)

File: app/domain/logs.rb [hunk 12]
```diff
@@ -590,6 +590,20 @@ module AuthnJwt
...
+
+      ValidatingClaimName = ::Util::TrackableLogMessageClass.new(
+        msg: "Validating claim name '{0-claim-name}'",
+        code: "CONJ00118D"
+      )
+
+      ValidatedClaimName = ::Util::TrackableLogMessageClass.new(
+        msg: "Successfully validated claim name '{0-claim-name}'",
+        code: "CONJ00119D"
+      )
```
Comment (hilagross) [inline app/domain/logs.rb:606]: same comment -
1. I'm not sure we the phrase 'claim name' is correct, I think it should be either claim or claim key.
2. denylist in one word -> see [this](https://ca-il-confluence.il.cyber-ark.com/display/rndp/%5BSpike%5D+Rename+and+offensive+of+conjur#id-[Spike]Renameandoffensiveofconjur-Specificchanges)

**Review by hilagross — CHANGES_REQUESTED** [review]

File: app/domain/authentication/authn_jwt/consts.rb (hunk 5 above)
Comment (hilagross) [reply app/domain/authentication/authn_jwt/consts.rb:32]: Agree with @sashaCher I would have add it to our documentation task now so we won't forget it.

File: app/domain/errors.rb (hunk 2 above)
Comment (hilagross) [reply app/domain/errors.rb:486]: @nessiLahav it looks that this wasn't changed.
I added my comment to the ones who did

File: app/domain/errors.rb (hunk 11 above)
Comment (hilagross) [reply app/domain/errors.rb:502]: @shulifink WDYT?

File: app/domain/logs.rb (hunk 12 above)
Comment (hilagross) [reply app/domain/logs.rb:606]: @shulifink WDYT?

File: app/domain/errors.rb (hunk 11 above)
Comment (sashaCher) [reply app/domain/errors.rb:502]: @hilagross "claim name" is looking like appropriate term see [RFC](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1).

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+          validate_claim_name_exists
+          validate_claim_name_value
+          validate_claim_is_allowed
+          @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatedClaimName.new(@claim_name))
+        end
+
+        private
+
+        def validate_claim_name_exists
+          raise Errors::Authentication::AuthnJwt::FailedToValidateClaimMissingClaimName if @claim_name.blank?
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:24]: i think that `MissingClaimName` is enough and much more readable

Comment (orenbm) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:24]: also - i think that `EmptyClaimName` is better

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        end
+
+        def validate_claim_name_value
+          return if claim_name_regex.match?(@claim_name)
+
+          raise Errors::Authentication::AuthnJwt::FailedToValidateClaimForbiddenClaimName.new(
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:30]: i think that `ForbiddenClaimName` is enough

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+            @claim_name,
+            MANDATORY_CLAIMS_RESOURCE_NAME,
+            claim_name_regex
+          )
+        end
+        
+        def claim_name_regex
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:37]: how about `valid_claim_name_regex`?

File: app/domain/errors.rb (hunk 11 above)
Comment (hilagross) [reply app/domain/errors.rb:502]: great, so we should keep it :)
still need to change denylist

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 16]
```diff
@@ -0,0 +1,55 @@
+module Authentication
+  module AuthnJwt
+    module InputValidation
+      # Validate the claim name value
+      ValidateClaimName ||= CommandClass.new(
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:5]: it's quite hard to review this class without seeing its application usage. i think that this PR should include the code that uses this class in the app code.

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        dependencies: {
+          regexp_class: Regexp,
+          deny_claims_list_value: [],
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:8]: will this be set as input in the app code or just in UTs?

File: app/domain/authentication/authn_jwt/consts.rb (hunk 5 above)
Comment (benoaviram) [reply app/domain/authentication/authn_jwt/consts.rb:32]: @sashaCher why adding aud? if the customer will want to enforce aud only to one instance of authenticator? (without using the variable)

Comment (sashaCher) [reply app/domain/authentication/authn_jwt/consts.rb:32]: As long as `aud` is enforced on authenticator's level (like `iss` ) adding it to mandatory and/or mapping is useless.

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb (hunk 10 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:32]: Great comment, i will review again this class from the thinking that other component will use it

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb (hunk 17 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:8]: In app code

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb (hunk 16 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:5]: Correct, but i wanted to create small PRs

File: app/domain/errors.rb (hunk 11 above)
Comment (nessiLahav) [reply app/domain/errors.rb]: Done

File: app/domain/logs.rb (hunk 12 above)
Comment (nessiLahav) [reply app/domain/logs.rb]: Done

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb (hunk 13 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:24]: I think it is to general name and we may need it in the future in other flow

File: app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb (hunk 14 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/input_validation/validate_claim_name.rb:30]: same reason: I think it is to general name and we may need it in the future in other flow

**Review by hilagross — DISMISSED** [review]

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+          claim_name: "$"
+        )
+      end
+
+      it 'does not raise error' do
+        expect { subject }.not_to raise_error
+      end
+    end
+
+    context "When claim name contains digits in the middle" do
```
Comment (benoaviram) [inline spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:160]: no digits in the middle

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+        )
+      end
+
+      it 'does not raise error' do
+        expect { subject }.not_to raise_error
+      end
+    end
+
+    context "When claim name ends with digits" do
+      subject do
```
Comment (benoaviram) [inline spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:173]: this is the previous test. no ending with digits scenario

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+
+      it 'does not raise error' do
+        expect { subject }.not_to raise_error
+      end
+    end
+
+    context "When claim name contains digits in the middle" do
+      subject do
+        ::Authentication::AuthnJwt::InputValidation::ValidateClaimName.new().call(
+          claim_name: "$"
```
Comment (sashaCher) [inline spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:163]: Where are digits :-0?

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+
+      it 'does not raise error' do
+        expect { subject }.not_to raise_error
+      end
+    end
+
+    context "When claim name ends with digits" do
+      subject do
+        ::Authentication::AuthnJwt::InputValidation::ValidateClaimName.new().call(
+          claim_name: "$2w"
```
Comment (sashaCher) [inline spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:175]: It does not end with digit :-0.

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+          claim_name: "exp"
+        )
+      end
+
+      it "raises an error" do
+        expect { subject }.to raise_error(Errors::Authentication::AuthnJwt::FailedToValidateClaimClaimNameInDenyList)
+      end
+    end
+
+    context "When claim name value is 'aud'" do
```
Comment (benoaviram) [inline spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:200]: Worth adding scenario for each forbidden claim name.. to make sure someone won't delete one in the future by mistake
Also, the same as tamir, I think worth adding a test that will not raise an error when the claim name is contain a substring of forbidden claim name. i.e. claim_name: "expref"

**Review by sashaCher — COMMENTED** [review]
Two tests are looking wrong

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb (hunk 18 above)
Comment (nessiLahav) [reply spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:160]: thanks fixed, the error was in the input not in the title

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb (hunk 19 above)
Comment (nessiLahav) [reply spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:173]: correct and fixed

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb (hunk 21 above)
Comment (nessiLahav) [reply spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:175]: fixed

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb (hunk 20 above)
Comment (nessiLahav) [reply spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:163]: fixed

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+        expect { subject }.not_to raise_error
+      end
+    end
+
+    context "When claim name value is substring of forbidden claim '$exp'" do
+      subject do
+        ::Authentication::AuthnJwt::InputValidation::ValidateClaimName.new(
+          deny_claims_list_value: ::Authentication::AuthnJwt::MANDATORY_CLAIMS_DENY_LIST
+        ).call(
+          claim_name: "sub"
```
Comment (benoaviram) [inline spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:291]: Wrong?

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+        expect { subject }.not_to raise_error
+      end
+    end
+
+    context "When claim name value is substring of forbidden claim 'exp1'" do
+      subject do
+        ::Authentication::AuthnJwt::InputValidation::ValidateClaimName.new(
+          deny_claims_list_value: ::Authentication::AuthnJwt::MANDATORY_CLAIMS_DENY_LIST
+        ).call(
+          claim_name: "sub"
```
Comment (benoaviram) [inline spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:277]: Wrong?

File: spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb [hunk 25]
```diff
@@ -0,0 +1,300 @@
...
+    end
+  end
+
+  context "Claim name is not exists in deny list" do
+    context "When claim name value is 'sub'" do
+      subject do
+        ::Authentication::AuthnJwt::InputValidation::ValidateClaimName.new(
+          deny_claims_list_value: ::Authentication::AuthnJwt::MANDATORY_CLAIMS_DENY_LIST
+        ).call(
+          claim_name: "sub"
```
Comment (benoaviram) [inline spec/app/domain/authentication/authn-jwt/input_validation/validate_claim_name_spec.rb:263]: Missing a case for iss

**Review by benoaviram — APPROVED** [review]
